### PR TITLE
Derive quantizer index offsets for segments from distortion scales

### PR DESCRIPTION
**Pull request by barrbrain**
Runs k-means on logarithms of distortion scale, with k chosen by variance in spacing of the result.
Now with more comments and AWCY results.

### CODE REVIEW

File: src/rate.rs [hunk 1]
```diff
@@ -229,7 +229,7 @@ pub(crate) fn bexp64(logq57: i64) -> i64 {
 // input: a 64-bit integer in Q0 (no fraction).
 // output: a 64-bit log in Q57.
 // TODO: Mark const once we can use local variables in a const function.
-fn blog64(w: i64) -> i64 {
+pub fn blog64(w: i64) -> i64 {
```
Comment (lu-zero) [inline src/rate.rs:232]: note to self, once it lands refactor the maths function from rate.rs to a stand alone file.

**Review by lu-zero — APPROVED** [review]

File: src/segmentation.rs [hunk 2]
```diff
@@ -24,34 +26,46 @@ pub fn segmentation_optimize<T: Pixel>(
...
+    let centroids = kmeans(64, &mut log_scales);
+    let qidx_diffs: ArrayVec<[_; 3]> = {
+      use crate::rate::{bexp64, blog64};
+      let log_ac_q =
+        blog64(ac_q(fi.base_q_idx, 0, fi.sequence.bit_depth) as i64);
+      centroids
+        .iter()
+        .map(|&c| bexp64(log_ac_q - ((c as i64) << 31)))
+        .map(|q| {
+          select_ac_qi(q, fi.sequence.bit_depth) as i16 - fi.base_q_idx as i16
```
Comment (barrbrain) [inline src/segmentation.rs:50]: Should add `.max(1)` here to enforce the no-lossless constraint.

**Comment by barrbrain** [comment]
[AWCY results at the default speed](https://beta.arewecompressedyet.com/?job=master-87c0297c631cf40bc1e9b03e8130928bd8e223b7&job=segment-opt%402021-03-17T09%3A28%3A36.965Z) indicate more work is needed from this initial version:
| PSNR Y | PSNR Cb | PSNR Cr | CIEDE2000 |   SSIM | MS-SSIM | PSNR-HVS Y | PSNR-HVS Cb | PSNR-HVS Cr | PSNR-HVS |   VMAF | VMAF-NEG |
|   ---: |    ---: |    ---: |      ---: |   ---: |    ---: |       ---: |        ---: |        ---: |     ---: |   ---: |     ---: |
| 1.6103 |  1.8378 |  1.7304 |    1.8383 | 1.5591 |  1.7372 |     1.8687 |      1.8090 |      1.4279 |   1.8676 | 2.1277 |   2.1061 |

**Comment by shssoichiro** [comment]
Wow, this seems much cleaner. Nice!

**Comment by barrbrain** [comment]
AWCY results with number of [segments fixed at 3 and dynamic offsets on objective-fast-1 at default speed](https://beta.arewecompressedyet.com/?job=master-537b1b91b861210062b1cf105a5bd0243bab0b90&job=segment-opt-3-seg%402022-08-17T03%3A25%3A30.728Z):
|  PSNR Y | PSNR Cb | PSNR Cr | CIEDE2000 |   SSIM | MS-SSIM | PSNR-HVS Y | PSNR-HVS Cb | PSNR-HVS Cr | PSNR-HVS |    VMAF | VMAF-NEG |
|    ---: |    ---: |    ---: |      ---: |   ---: |    ---: |       ---: |        ---: |        ---: |     ---: |    ---: |     ---: |
| -1.4059 | -1.8891 | -1.9159 |   -0.7324 | 0.6363 |  0.6321 |    -0.8216 |     -1.5665 |     -1.4454 |  -0.8571 | -1.1439 |  -1.2585 |

**Comment by shssoichiro** [comment]
I noticed the testing between different numbers of segments--the reasoning for the current version using a dynamic number of segments based on the variance of scores within the frame is because there is some amount of overhead in using a higher number of segments (because they have to be coded). Therefore, we wanted to avoid coding segments that were not useful.

If it's not possible to dynamically choose the segment count based on score variance, maybe a workaround could be to use 5 segments with low-to-mid quantizers, where the segmentation coding overhead is less significant, and 3 segments with higher quantizers where the segmentation coding would affect the overall filesize more.

**Comment by barrbrain** [comment]
The goal is to retain a dynamic number of segments. The current testing is to measure the performance of an 'oracle' that picks the best number of segments for each clip.

**Comment by barrbrain** [comment]
AWCY results for [objective-1-fast at default speed](https://beta.arewecompressedyet.com/?job=master-1816a1c625b9279369d868328f6a2e801637ea43&job=segment-opt%402022-08-21T08%3A41%3A58.294Z)

|  PSNR Y | PSNR Cb | PSNR Cr | CIEDE2000 |    SSIM | MS-SSIM | PSNR-HVS Y | PSNR-HVS Cb | PSNR-HVS Cr | PSNR-HVS |    VMAF | VMAF-NEG |
|    ---: |    ---: |    ---: |      ---: |    ---: |    ---: |       ---: |        ---: |        ---: |     ---: |    ---: |     ---: |
| -0.4139 | -0.1869 | -0.2631 |   -0.2345 | -0.2842 | -0.2500 |    -0.3600 |      0.1368 |     -0.2013 |  -0.3594 | -0.3614 |  -0.4027 |

**Review by lu-zero — APPROVED** [review]

**Pull request merged**